feat: add IgnoreObsoleteMembers and `MapProperty` maps ignored Target

TimothyMakkison opened the pull request:
Resolves #284 and #450.

- Created `MapperIgnoreObsoleteMembersAttribute` and added `IgnoreObsoleteStrategy` to `MapperAttribute`. Updated `MembersMappingBuilderContext` to add the derecated members to there respective `Ignored*MemberNames`.
- Updated `FieldMember` and `PropertyMember` to expose the underlying symbol.
- Added unit tests
- Mapper attribute config is overriden by the method configuration.
- By default `IgnoreObsoleteStrategy` is none so deprecated members are treated as normal.
- The method attribute has a default parameters of `Both`  
```C#
[MapperIgnoreObsoleteMembers] partial B Map(A source);`
```

>  blanket ignore all obsolete members except when [MapProperty] is used?

Haven't done this yet, I'm not 100% sure how `MemberConfigsByRootTargetName` works to match the obsolete members to the respective explicitly mapped member.

I'll wait for #390 and then update the relevant logic.

TimothyMakkison commented:
Does Mapperly support ignoring nested properties?

```C#
// Where ProducerDto has getter setter methods and a parameterless constructor.
[MapperIgnoreSource("Manufacturer.Id")]
[MapperIgnoreTarget("Producer.Id")]
[MapProperty(nameof(Car.Manufacturer), nameof(CarDto.Producer))] // Map property with a different name in the target type
public static partial CarDto MapCarToDto(Car car);
```
### Generates
```C#
 public static partial global::Riok.Mapperly.Sample.CarDto MapCarToDto(global::Riok.Mapperly.Sample.Car car)
{
    var target = new global::Riok.Mapperly.Sample.CarDto();
    if (car.Manufacturer != null)
    {
        target.Producer = MapToProducerDto(car.Manufacturer);
    }

    target.Name = car.Name;
//...
private static global::Riok.Mapperly.Sample.ProducerDto MapToProducerDto(global::Riok.Mapperly.Sample.Manufacturer source)
{
    var target = new global::Riok.Mapperly.Sample.ProducerDto();
    target.Id = source.Id;
    target.Name = source.Name;
    return target;
}
```

latonz commented:
@TimothyMakkison nested ignored properties are currently not supported. Is there a use-case for this?

latonz commented on test/Riok.Mapperly.Tests/Mapping/IgnoreObsoleteTest.cs:
```diff
@@ -0,0 +1,227 @@
+
+public class IgnoreObsoleteTest
+{
+    private readonly string _classA = """
```
Couldn't this be a const?

latonz commented on test/Riok.Mapperly.Tests/Mapping/IgnoreObsoleteTest.cs:
```diff
@@ -0,0 +1,227 @@
+    [Fact]
+    public void MethodAttributeIgnoreObsoleteBoth()
+    {
+        var source = TestSourceBuilder.MapperWithBodyAndTypes("[MapperIgnoreObsoleteMembers()] partial B Map(A source);", _classA, _classB);
```
Nit: rm the unneeded parenthesis

latonz commented on test/Riok.Mapperly.Tests/Mapping/IgnoreObsoleteTest.cs:
```diff
@@ -0,0 +1,227 @@
+        """;
+
+    [Fact]
+    public void ClassAttributeIgnoreObsoleteNone()
```
Add a test case which explicitly maps a obsolete property via `[MapProperty]` which should overwrite the obsolete member strategy.

TimothyMakkison commented:
> @TimothyMakkison nested ignored properties are currently not supported. Is there a use-case for this?

Mb I thought it did, wasn't sure if i had messed up the member paths. I wanted to reuse some of the logic to let explicit mappings override ignored deprecated members.

I think nested ignores would be a good addition to mapperl, It already has nested explicit mappings.

latonz reviewed (changes requested):
Thank you for another great contribution 🌞 
* Good looking tests 😊
* Could you also update the documentation?
* TBD: Should obsolete enum members also be ignored (to be discussed, if implemented, implement it in another PR). Should it be automatically or via a separate configuration? I think the best way would be to do it the exact same way, without any additional configuration. If it should be overwritten for an enum, it could get overwritten on an enum mapping method with the same attribute. WDYT?

latonz commented:
@TimothyMakkison 
> I think nested ignores would be a good addition to mapperl, It already has nested explicit mappings.

Can you show a concrete use-case for this? Probably open a new discussion about this topic to not bloat this PR.

latonz commented on src/Riok.Mapperly.Abstractions/MapperIgnoreObsoleteMembersAttribute.cs:
```diff
@@ -0,0 +1,15 @@
+/// <summary>
+/// Specifies options for obsolete ignoring strategy.
+/// </summary>
+[AttributeUsage(AttributeTargets.Method, AllowMultiple = false)]
```
`AllowMultiple` can be removed as it is the default anyway.

latonz commented:
@TimothyMakkison FYI: an issue to the nested ignored properties just got opened: https://github.com/riok/mapperly/issues/395

latonz commented:
@TimothyMakkison feel free to re-request a review when ready 😊

TimothyMakkison commented:
> ignore all obsolete members except when [MapProperty] is used?


Discovered #450 when trying to implement this. I could fix the bug so `MapProperty` will map ignored target properties just like obsolete target properties.

TimothyMakkison commented on src/Riok.Mapperly/Symbols/IMappableMember.cs:
```diff
@@ -9,6 +9,8 @@ public interface IMappableMember
 
     ITypeSymbol Type { get; }
 
+    ISymbol MemberSymbol { get; }
```
I originally didn't do this because:
- This property is only used when `IgnoreObsolete` is enabled and could slow down generation.
- Requires a `WellKnownTypes` instance being passed to `MappableMember.Create` / `new PropertyMember`.

WDYT?

latonz replied:
I think your points make sense, exposing it is okay with me.

TimothyMakkison commented:
- `MapProperty` now always maps ignored obsolete members.
  - In the process resolved #450 decided to make it so `MapProperty` overrides ingored target properties just like it already did for ignored source proeprties.
- Updated `PublicAPI.Shipped` file.
- Added `MapProperty` tests.
- I haven't added to the docs, should I create a brief summary, and if so where?

latonz commented on src/Riok.Mapperly/Helpers/DictionaryExtensions.cs:
```diff
@@ -18,4 +18,19 @@ public static class DictionaryExtensions
         }
     }
+
+    public static Dictionary<TKey, TValue> ExtractRange<TKey, TValue>(this IDictionary<TKey, TValue> dict, IEnumerable<TKey> keys)
```
Could you unit test this?

latonz commented on src/Riok.Mapperly/Descriptors/MappingBodyBuilders/BuilderContext/MembersMappingBuilderContext.cs:
```diff
@@ -66,6 +75,40 @@ private HashSet<string> InitIgnoredUnmatchedProperties(IEnumerable<string> allPr
+    {
+        var obsoleteStrategy = BuilderContext.Configuration.Properties.IgnoreObsoleteMembersStrategy;
+
+        if (obsoleteStrategy.HasFlag(IgnoreObsoleteMembersStrategy.Target))
+        {
+            var obsoleteAttribute = BuilderContext.Types.Get<ObsoleteAttribute>();
+
+            return Mapping.TargetType
+                .GetAccessibleMappableMembers()
+                .Where(x => x.MemberSymbol.HasAttribute(obsoleteAttribute))
+                .Select(x => x.Name);
+        }
+
+        return Enumerable.Empty<string>();
```
I'd invert the if condition, same for `GetIgnoredObsoleteSourceMembers` (IMO improves the readability).

latonz commented on src/Riok.Mapperly/Descriptors/MappingBodyBuilders/NewInstanceObjectMemberMappingBodyBuilder.cs:
```diff
@@ -66,6 +66,19 @@ private static void BuildInitOnlyMemberMappings(INewInstanceBuilderContext<IMapp
             BuildInitMemberMapping(ctx, targetMember, sourceMemberPath);
         }
+
+        var initOnlyIgnoredTargetMembers = includeAllMembers
+            ? ctx.IgnoredTargetMembers.Values.ToArray()
+            : ctx.IgnoredTargetMembers.Values.Where(x => x.CanOnlySetViaInitializer()).ToArray();
+        foreach (var targetMember in initOnlyIgnoredTargetMembers)
+        {
+            ctx.IgnoredTargetMembers.Remove(targetMember.Name);
+
+            if (ctx.MemberConfigsByRootTargetName.Remove(targetMember.Name, out var memberConfigs))
+            {
+                BuildInitMemberMapping(ctx, targetMember, memberConfigs);
+            }
+        }
```
Could you add a comment what this does and why it is needed? Does this work nicely with required init properties?

TimothyMakkison replied:
Added comment and added required/init tests

latonz commented on src/Riok.Mapperly/Descriptors/MappingBodyBuilders/ObjectMemberMappingBodyBuilder.cs:
```diff
@@ -66,6 +66,20 @@ public static void BuildMappingBody(IMembersContainerBuilderContext<IMemberAssig
+        {
+            if (ctx.MemberConfigsByRootTargetName.Remove(targetMember.Name, out var memberConfigs))
+            {
+                // add all configured mappings
```
probably add ` even if they are ignored` and move it up to L69 as this is where the section begins IMO.

latonz reviewed (changes requested):
Thank you for the updates, I added my feedback.
Could you also update the documentation (I would add a new section in `01-mapper.md` in `## Properties / fields`).

TimothyMakkison commented:
Note that in order to support `MapProperty` mapping obsolete targets, I fixed #450 in this PR. 

Now if a target is Ignored but is explicitly mapped with `MapProperty` a mapping will be created.

latonz commented on src/Riok.Mapperly/Descriptors/MappingBodyBuilders/NewInstanceObjectMemberMappingBodyBuilder.cs:
```diff
@@ -66,6 +66,19 @@ private static void BuildInitOnlyMemberMappings(INewInstanceBuilderContext<IMapp
 
             BuildInitMemberMapping(ctx, targetMember, sourceMemberPath);
         }
+
+        // check for any configured mappings that match the ignored targets
+        // this way explicit property mapping will map ignored members
+        var initOnlyIgnoredTargetMembers = includeAllMembers
+            ? ctx.IgnoredTargetMembers.Values.ToArray()
+            : ctx.IgnoredTargetMembers.Values.Where(x => x.CanOnlySetViaInitializer()).ToArray();
+        foreach (var targetMember in initOnlyIgnoredTargetMembers)
+        {
+            if (ctx.MemberConfigsByRootTargetName.Remove(targetMember.Name, out var memberConfigs))
+            {
+                BuildInitMemberMapping(ctx, targetMember, memberConfigs);
+            }
+        }
```
Instead of this, wouldn't it be much simpler to just add `ignoredTargetMemberNames.ExceptWith(MemberConfigsByRootTargetName.Keys);` in `MembersMappingBuilderContext` before initialising `IgnoredTargetMembers` (L45)?

TimothyMakkison replied:
It would save time by removing ignored targets that are not explicitly mapped. I don't think it would replace this logic tho

latonz reviewed (changes requested):
I'd add a warning diagnostic if an explicit ignored property is mapped manually (but not if a property is ignored because it is obsolete). I think this could be done in another PR.

Could you adjust the PR body that it resolves https://github.com/riok/mapperly/issues/450?

latonz commented on src/Riok.Mapperly/Descriptors/MappingBodyBuilders/BuilderContext/MembersMappingBuilderContext.cs:
```diff
@@ -66,6 +76,37 @@ private HashSet<string> InitIgnoredUnmatchedProperties(IEnumerable<string> allPr
+    {
+        var obsoleteStrategy = BuilderContext.Configuration.Properties.IgnoreObsoleteMembersStrategy;
+
+        if (obsoleteStrategy.HasFlag(IgnoreObsoleteMembersStrategy.Source))
```
I'd invert this condition to improve the readability. Similar to the style which is implemented by `GetIgnoredObsoleteTargetMembers`.

latonz reviewed (changes requested):
@TimothyMakkison I created https://github.com/riok/mapperly/issues/512 to implement the diagnostic.
Whats still missing:
* Update the documentation (I would add a new section in `01-mapper.md` in `## Properties / fields`)
* Update integration tests

Otherwise this MR looks good 😊

latonz commented:
@TimothyMakkison looks good to me. A few linter issues need to be fixed...

TimothyMakkison commented:
> @TimothyMakkison looks good to me. A few linter issues need to be fixed...

Thanks, please don't merge this just yet, I haven't added the integration tests

latonz commented:
Oh thanks, I would have missed that 🙈
